core/isolate.rs

```diff
@@ -17,13 +17,52 @@ use libc::c_void;
 use std::collections::VecDeque;
 use std::ffi::CStr;
 use std::ffi::CString;
+use std::fmt::{Debug, Formatter, Result as FmtResult};
 use std::sync::{Arc, Mutex, Once, ONCE_INIT};
 
 pub type Buf = Box<[u8]>;
-pub type Op = dyn Future<Item = Buf, Error = ()> + Send;
+
+pub trait Op: Future + Debug + Sized + Send {
+  fn new<F: Future<Item = Self::Item, Error = Self::Error> + Send + 'static>(
+    label: &'static str,
+    future: F,
+  ) -> Self;
+}
+
+pub struct BoxOp<I, E> {
+  pub label: &'static str,
```

Comment:
This might raise some eyebrows but seems fine after I thought about it for a second. We statically branch on ops. I don't think we need to enable people to have dynamic ops.